cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -299,385 +552,988 @@ class RleEncoder {
+      validity_run = validity_reader.NextRun();
     }
+  }
+
+  value_type const value = decoder.Value();
+  if (ARROW_PREDICT_FALSE(!converter.InputIsValid(value))) {
+    return {0, 0};
+  }
+  converter.WriteRepeated(out, out + batch.TotalRead(), value);
+  auto const actual_values_read = decoder.Advance(batch.ValuesRead());
+  // We always cropped the number of values_read by the remaining values in the run.
+  // What's more the RLE decoder should not encounter any errors.
+  ARROW_DCHECK_EQ(actual_values_read, batch.ValuesRead());
 
-    ARROW_DCHECK_GT(batch_size, 0);
-    ARROW_DCHECK_GT(valid_run.length, 0);
+  return {batch.ValuesRead(), batch.NullRead()};
+}
+
+template <typename T, typename... Ts>
+[[nodiscard]] constexpr T min(T x, Ts... ys) {
+  ((x = std::min(x, ys)), ...);
+  return x;
+}
+
+static_assert(min(5) == 5);
+static_assert(min(5, 4, -1) == -1);
+static_assert(min(5, 41) == 5);
+
+template <typename Converter, typename BitRunReader, typename BitRun,
+          typename values_count_type, typename value_type>
+auto RunGetSpaced(Converter& converter, typename Converter::out_type* out,
+                  values_count_type batch_size, values_count_type null_count,
+                  BitRunReader&& validity_reader, BitRun&& validity_run,
+                  BitPackedDecoder<value_type>& decoder)
+    -> std::pair<values_count_type, values_count_type> {
+  ARROW_DCHECK_GT(batch_size, 0);
+  // The equality case is handled in the main loop in GetSpaced
+  ARROW_DCHECK_LT(null_count, batch_size);
+
+  auto batch = BatchCounter::FromBatchSizeAndNulls(batch_size, null_count);
+
+  values_count_type const values_available = decoder.Remaining();
+  ARROW_DCHECK_GT(values_available, 0);
+  auto run_values_remaining = [&]() {
+    auto out = values_available - batch.ValuesRead();
+    ARROW_DCHECK_GE(out, 0);
+    return out;
+  };
+
+  while (run_values_remaining() > 0 && batch.ValuesRemaining() > 0) {
+    // TODO should this size be tune depending on sizeof(value_size)? cpu cache size?
+    // Pull a batch of values from the bit packed encoded data and store it in a local
+    // buffer to benefit from unpacking intrinsics and data locality.
+    static constexpr values_count_type kBufferCapacity = 1024;
+    std::array<value_type, kBufferCapacity> buffer = {};
+
+    values_count_type buffer_start = 0;
+    values_count_type buffer_end = 0;
+    auto buffer_size = [&]() {
+      auto out = buffer_end - buffer_start;
+      ARROW_DCHECK_GE(out, 0);
+      return out;
+    };
+
+    // buffer_start is 0 at this point so size is end
+    buffer_end = min(run_values_remaining(), batch.ValuesRemaining(), kBufferCapacity);
+    buffer_end = decoder.GetBatch(buffer.data(), buffer_size());
+    ARROW_DCHECK_LE(buffer_size(), kBufferCapacity);
+
+    if (ARROW_PREDICT_FALSE(!converter.InputIsValid(buffer.data(), buffer_size()))) {
+      return {batch.ValuesRead(), batch.NullRead()};
```

Comment:
The same error case for `RleDecoder` returns `{0, 0}`, why do we return a partial result instead here.

Reply:
If easily doable then yes.